boa_engine/src/object/jsmap.rs

```diff
@@ -0,0 +1,178 @@
+// This module is a wrapper for the Map Builtin Javascript Object
+use crate::{
+    builtins::map::{add_entries_from_iterable, ordered_map::OrderedMap},
+    builtins::Map,
+    object::{JsFunction, JsMapIterator, JsObject, JsObjectType, ObjectData},
+    Context, JsResult, JsValue,
+};
+
+use boa_gc::{Finalize, Trace};
+use std::ops::Deref;
+
+#[derive(Debug, Clone, Trace, Finalize)]
+pub struct JsMap {
+    inner: JsObject,
+}
+
+impl JsMap {
+    /// Create new Empty Map Object.
+    #[inline]
+    pub fn new(context: &mut Context) -> Self {
+        let map = Self::create_map(context);
+        Self { inner: map }
+    }
+
+    /// Create a new map object for any object that has a `@@Iterator` field.
+    #[inline]
+    pub fn from_js_iterable(iterable: &JsValue, context: &mut Context) -> JsResult<Self> {
+        // Create a new map object.
+        let map = Self::create_map(context);
+
+        // Let adder be Get(map, "set") per spec. This action should not fail with default map.
+        let adder = map
+            .get("set", context)
+            .expect("creating a map with the default prototype must not fail");
+
+        let _completion_record = add_entries_from_iterable(&map, iterable, &adder, context)?;
+
+        Ok(Self { inner: map })
+    }
+
+    /// Create a [`JsMap`] from a [`JsObject`], if the object is not a `Map`, throw a `TypeError`.
+    #[inline]
+    pub fn from_object(object: JsObject, context: &mut Context) -> JsResult<Self> {
+        if object.borrow().is_map() {
+            Ok(Self { inner: object })
+        } else {
+            context.throw_type_error("object is not a Map")
+        }
+    }
+
+    // Utility function to generate default Map object.
+    fn create_map(context: &mut Context) -> JsObject {
+        // Get default Map prototype
+        let prototype = context.intrinsics().constructors().map().prototype();
+
+        // Create a default map object with [[MapData]] as a new empty list
+        JsObject::from_proto_and_data(prototype, ObjectData::map(OrderedMap::new()))
+    }
```

Comment:
I mostly used this for creating the default in `new` and `from_js_iterable`. I also figured it may be useful if there was ever the desire in the future to create more "from" methods. For instance, if there is a desire to create a `from_iter` method that takes a rust iterator and creates a map. It can be split into `new` and `from_js_iterable`, but would it be worth saving for now?

Reply:
Hm, it can be left as it is then :)